src/ptvsd/adapter/server.py

```diff
@@ -141,4 +261,116 @@ def exited_event(self, event):
     @message_handler
     def terminated_event(self, event):
         # Do not propagate this, since we'll report our own.
-        pass
+        self.channel.close()
+
+    def detach_from_session(self):
+        with _lock:
+            self.is_connected = False
+            self.channel.handlers = self.connection
+            self.channel.name = self.channel.stream.name = str(self.connection)
+            self.connection.server = None
+
+    def disconnect(self):
+        with _lock:
+            _connections.remove(self.connection)
+            _connections_changed.set()
+        super(Server, self).disconnect()
+
+
+listen = Connection.listen
+
+
+def connections():
+    with _lock:
+        return list(_connections)
+
+
+def wait_for_connection(pid=any, timeout=None):
+    """Waits until there is a server with the specified PID connected to this adapter,
+    and returns the corresponding Connection.
+
+    If there is more than one server connection already available, returns the oldest
+    one.
+    """
+
+    def wait_for_timeout():
+        time.sleep(timeout)
+        wait_for_timeout.timed_out = True
+        with _lock:
```

Comment:
Minor: I don't think you need the lock to set the event.

Reply:
On a 2nd thought, maybe we do (and we should also add the `with _lock` to `dont_expect_connections`).